### DISCUSSION TITLE
loop cache performance

### DISCUSSION THREAD

**Discussion by lagden** [frontendbr/forum#271] [Dicas]
Hoje esta lendo sobre dicas de javascript e estava falando de cachear o `length` do `array` antes de chamar o loop `for`...

``` js
for (var i = 0, len = arr.length; i < len; i++) {
    arr[i]
}
```

Logo abaixo nos comentários um rapaz disse: `Bullshit`
Fiquei na dúvida e fiz um benchmark!!

E não era que o sujeito estava certo!!

https://gist.github.com/lagden/0ec5360c39e9a74783d6071a96c8c59e

![screen shot 2016-08-01 at 17 55 59](https://cloud.githubusercontent.com/assets/130963/17308806/2f848204-5812-11e6-8720-2063922e2410.png)

`for1.js` com cache
`for2.js` sem cache

Ou seja, com ou sem da na mesma!!
O que me dizem?

**Comment by fernandofleury** [comment]
Essa dica é antiga. Hoje as engines de JS já cuidam dessa otimização pra gente :)

**Comment by lagden** [comment]
@fernandofleury Bacana!! Então se rodar no IE 8 ou 9 vai rolar uma diferença?

**Comment by fernandofleury** [comment]
@lagden No 9 eu não sei, mas no 8 com certeza.

**Comment by lagden** [comment]
@fernandofleury testei no IE 11

É agressivo mesmo!!! Quase a metade do tempo!!

``` js
// IE com cache
function for1IE() {
    'use strict';

    var a = [];
    for (var _i = 0; _i < 9999999; _i++) {
        a.push(_i);
    }

    var before = new Date();
    for (var c = 0; c < 1e3; c++) {
        // ...
    }

    for (var _i2 = 0, _len = a.length; _i2 < _len; _i2++) {
        // ...
    }
    var after = new Date();
    console.log(after.getTime() - before.getTime());
}

for1IE();
```

``` js
// IE sem cache
function for2IE() {
    'use strict';

    var a = [];
    for (var _i = 0; _i < 9999999; _i++) {
        a.push(_i);
    }

    var before = new Date();
    for (var c = 0; c < 1e3; c++) {
        // ...
    }

    for (var _i2 = 0; _i2 < a.length; _i2++) {
        // ...
    }
    var after = new Date();
    console.log(after.getTime() - before.getTime());
}

for2IE();
```

**Comment by eduardojmatos** [comment]
@lagden coloca o resultado do teste com IE11 pra gente!! =D
